aiida/tools/importexport/archive/common.py

```diff
@@ -61,3 +78,302 @@ def detect_archive_type(in_path: str) -> str:
+            for membername in members:
+                callback('update', 1)
+                handle.extract(path=os.path.abspath(out_path), member=membername)
+    except zipfile.BadZipfile as error:
+        raise CorruptArchive(f'The input file cannot be read: {error}')
+
+
+def safe_extract_tar(
+    in_path: Union[str, Path],
+    out_path: Union[str, Path],
+    *,
+    only_prefix: Iterable[str] = (),
+    ignore_prefix: Iterable[str] = ('..', '/'),
+    callback: Callable[[str, Any], None] = null_callback,
+    callback_description: str = 'Extracting tar files'
+):
+    """Safely extract a tar file
+
+    :param in_path: Path to extract from
+    :param out_path: Path to extract to
+    :param only_prefix: Extract only internal paths starting with these prefixes
+    :param ignore_prefix: Ignore internal paths starting with these prefixes
+    :param callback: a callback to report on the process, ``callback(action, value)``,
+        with the following callback signatures:
+
+        - ``callback('init', {'total': <int>, 'description': <str>})``,
+            to signal the start of a process, its total iterations and description
+        - ``callback('update', <int>)``,
+            to signal an update to the process and the number of iterations to progress
+
+    :param callback_description: the description to return in the callback
+
+    :raises `~aiida.tools.importexport.common.exceptions.CorruptArchive`: if the file cannot be read
+
+    """
+    _filter = _get_filter(only_prefix, ignore_prefix)
+    try:
+        with tarfile.open(in_path, 'r:*', format=tarfile.PAX_FORMAT) as handle:
+            members = [m for m in handle.getmembers() if _filter(m.name)]
+            callback('init', {'total': len(members), 'description': callback_description})
+            for member in members:
+                callback('update', 1)
+                if member.isdev():
+                    # safety: skip if character device, block device or FIFO
+                    msg = f'WARNING, device found inside the tar file: {member.name}'
+                    ARCHIVE_LOGGER.warning(msg)
+                    continue
+                if member.issym() or member.islnk():
+                    # safety: skip symlinks
+                    msg = f'WARNING, symlink found inside the tar file: {member.name}'
+                    ARCHIVE_LOGGER.warning(msg)
+                    continue
+                handle.extract(path=os.path.abspath(out_path), member=member.name)
+    except tarfile.ReadError as error:
+        raise CorruptArchive(f'The input file cannot be read: {error}')
+
+
+class CacheFolder:
+    """A class to encapsulate a folder path with cached read/writes.
+
+    The class can be used as a context manager, and will flush the cache on exit::
+
+        with CacheFolder(path) as folder:
+            # this stored in memory (no disk write)
+            folder.write_text('path/to/file.txt', 'content')
```

Comment:
Could you add a json example here as well?

Reply:
done